src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenHandlerBase.cs

```diff
@@ -205,10 +223,28 @@ private void StoreResultExToCache(ref bool notifiedBeforeAccessCache)
                     this.NotifyBeforeAccessCache();
                     notifiedBeforeAccessCache = true;
                 }
-
-                this.tokenCache.StoreToCache(ResultEx, this.Authenticator.Authority, this.Resource,
+                var metadata = await InstanceDiscovery.GetMetadataEntry(
+                    GetHost(this.Authenticator.Authority), this.Authenticator.ValidateAuthority, this.CallState).ConfigureAwait(false);
+                this.tokenCache.StoreToCache(
+                    ResultEx, ReplaceHost(this.Authenticator.Authority, metadata.PreferredCache), this.Resource,
                     this.ClientKey.ClientId, this.TokenSubjectType, this.CallState);
             }
+            return notifiedBeforeAccessCache;
+        }
+
+        private string GetHost(string uri)
+        {
+            new Uri(uri); // This line serves as a validation for uri. Relevant exceptions will be throwed.
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either remove this useless object instantiation of class 'Uri' or use it. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS1848)